File: mobile/src/main/java/com/windscribe/mobile/ticket/SendTicketPresenterImpl.java
/*
 * Copyright (c) 2021 Windscribe Limited.
 */

package com.windscribe.mobile.ticket;

import android.content.Context;
import android.util.Patterns;

import com.windscribe.mobile.R;
import com.windscribe.vpn.ActivityInteractor;
import com.windscribe.vpn.api.CreateHashMap;
import com.windscribe.vpn.api.response.ApiErrorResponse;
import com.windscribe.vpn.api.response.GenericResponseClass;
import com.windscribe.vpn.api.response.QueryType;
import com.windscribe.vpn.api.response.TicketResponse;
import com.windscribe.vpn.constants.PreferencesKeyConstants;
import com.windscribe.vpn.repository.UserRepository;

import java.util.Map;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.observers.DisposableSingleObserver;
import io.reactivex.schedulers.Schedulers;

public class SendTicketPresenterImpl implements SendTicketPresenter {

    private QueryType queryType = QueryType.Account;

    private final ActivityInteractor sendTicketInteractor;

    private final SendTicketView sendTicketView;

    @Inject
    public SendTicketPresenterImpl(SendTicketView sendTicketView, ActivityInteractor activityInteractor) {
        this.sendTicketView = sendTicketView;
        this.sendTicketInteractor = activityInteractor;
    }

    @Override
    public void init() {
        sendTicketView.setActivityTitle(sendTicketInteractor.getResourceString(R.string.send_ticket));
        sendTicketView.setQueryTypeSpinner();
        sendTicketView.addTextChangeListener();
        if(sendTicketInteractor.getUserRepository().getUser().getValue()!=null && sendTicketInteractor.getUserRepository().getUser().getValue().getEmail()!=null){
            sendTicketView.setEmail(sendTicketInteractor.getUserRepository().getUser().getValue().getEmail());
        }
    }

    @Override
    public void onInputChanged(String email, String subject, String message) {
        sendTicketView.setSendButtonState(validEmail(email) && validMessage(message) && validSubject(subject));
    }

    @Override
    public void onQueryTypeSelected(QueryType queryType) {
        this.queryType = queryType;
    }

    @Override
    public void onSendTicketClicked(String email, String subject, String message) {
        sendTicketView.setProgressView(true);
        String username = sendTicketInteractor.getAppPreferenceInterface().getUserName();
        Map<String, String> queryMap = CreateHashMap.INSTANCE.buildTicketMap(email, subject, message, username, queryType);
        sendTicketInteractor.getCompositeDisposable()
                .add(sendTicketInteractor.getApiCallManager().sendTicket(queryMap)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(
                                new DisposableSingleObserver<GenericResponseClass<TicketResponse, ApiErrorResponse>>() {
                                    @Override
                                    public void onError(@NonNull Throwable e) {
                                        sendTicketView.setProgressView(false);
                                        sendTicketView.setErrorLayout("Failed to submit ticket. Try again.");
                                    }

                                    @Override
                                    public void onSuccess(
                                            @NonNull GenericResponseClass<TicketResponse, ApiErrorResponse> response) {
                                        sendTicketView.setProgressView(false);
                                        if (response.getDataClass() != null) {
                                            sendTicketView.setSuccessLayout(
                                                    "Sweet, we’ll get back to you as soon as one of our agents is back from lunch.");
                                        } else if (response.getErrorClass() != null) {
                                            sendTicketView.setErrorLayout(response.getErrorClass().getErrorMessage());
                                        } else {
                                            sendTicketView.setErrorLayout("Failed to submit ticket. Try again.");
                                        }
                                    }
                                }));
    }

    @Override
    public void setTheme(Context context) {
        String savedThem = sendTicketInteractor.getAppPreferenceInterface().getSelectedTheme();
        if (savedThem.equals(PreferencesKeyConstants.DARK_THEME)) {
            context.setTheme(R.style.DarkTheme);
        } else {
            context.setTheme(R.style.LightTheme);
        }
    }

    private boolean validEmail(String email) {
        return !email.isEmpty() && Patterns.EMAIL_ADDRESS.matcher(email).matches();
    }

    private boolean validMessage(String message) {
        return !message.isEmpty();
    }

    private boolean validSubject(String subject) {
        return !subject.isEmpty();
    }
}

File: mobile/src/main/java/com/windscribe/mobile/ticket/SendTicketPresenterImpl.kt
/*
 * Copyright (c) 2021 Windscribe Limited.
 */
package com.windscribe.mobile.ticket

import android.content.Context
import android.util.Patterns
import com.windscribe.mobile.R
import com.windscribe.vpn.ActivityInteractor
import com.windscribe.vpn.api.CreateHashMap.buildTicketMap
import com.windscribe.vpn.api.response.ApiErrorResponse
import com.windscribe.vpn.api.response.GenericResponseClass
import com.windscribe.vpn.api.response.QueryType
import com.windscribe.vpn.api.response.TicketResponse
import com.windscribe.vpn.constants.NetworkErrorCodes
import com.windscribe.vpn.constants.PreferencesKeyConstants
import com.windscribe.vpn.repository.CallResult
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.observers.DisposableSingleObserver
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

class SendTicketPresenterImpl @Inject constructor(
    private val sendTicketView: SendTicketView,
    private val interactor: ActivityInteractor
) : SendTicketPresenter {
    private var queryType = QueryType.Account
    override fun init() {
        sendTicketView.setActivityTitle(interactor.getResourceString(R.string.contact_humans))
        sendTicketView.setQueryTypeSpinner()
        sendTicketView.addTextChangeListener()
        interactor.getUserRepository().user.value?.email?.let {
            sendTicketView.setEmail(it)
        }
    }

    override fun onInputChanged(email: String, subject: String, message: String) {
        sendTicketView.setSendButtonState(
            validEmail(email) && validMessage(message) && validSubject(
                subject
            )
        )
    }

    override fun onQueryTypeSelected(queryType: QueryType) {
        this.queryType = queryType
    }

    override fun onSendTicketClicked(email: String, subject: String, message: String) {
        sendTicketView.setProgressView(true)
        val username = interactor.getAppPreferenceInterface().userName
        val queryMap = buildTicketMap(email, subject, message, username, queryType)
        interactor.getCompositeDisposable()
            .add(
                interactor.getApiCallManager().sendTicket(queryMap)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeWith(
                        object :
                            DisposableSingleObserver<GenericResponseClass<TicketResponse?, ApiErrorResponse?>>() {
                            override fun onError(e: Throwable) {
                                sendTicketView.setProgressView(false)
                                sendTicketView.setErrorLayout("Failed to submit ticket. Try again.")
                            }

                            override fun onSuccess(
                                response: GenericResponseClass<TicketResponse?, ApiErrorResponse?>
                            ) {
                                sendTicketView.setProgressView(false)
                                when (val result = response.callResult<TicketResponse>()) {
                                    is CallResult.Error -> {
                                        if (result.code == NetworkErrorCodes.ERROR_UNEXPECTED_API_DATA) {
                                            sendTicketView.setErrorLayout("Failed to submit ticket. Try again.")
                                        } else {
                                            sendTicketView.setErrorLayout(result.errorMessage)
                                        }
                                    }
                                    is CallResult.Success -> sendTicketView.setSuccessLayout("Sweet, we’ll get back to you as soon as one of our agents is back from lunch.")
                                }
                            }
                        })
            )
    }

    override fun setTheme(context: Context) {
        val savedThem = interactor.getAppPreferenceInterface().selectedTheme
        if (savedThem == PreferencesKeyConstants.DARK_THEME) {
            context.setTheme(R.style.DarkTheme)
        } else {
            context.setTheme(R.style.LightTheme)
        }
    }

    private fun validEmail(email: String): Boolean {
        return email.isNotEmpty() && Patterns.EMAIL_ADDRESS.matcher(email).matches()
    }

    private fun validMessage(message: String): Boolean {
        return message.isNotEmpty()
    }

    private fun validSubject(subject: String): Boolean {
        return subject.isNotEmpty()
    }
}